### DISCUSSION TITLE
Handling time windows when including service/duration times at nodes.

### DISCUSSION THREAD

**Discussion by nicomiguel6** [google/or-tools#4315] [Routing (and legacy CP) questions]
Hello,

I've browsed through a few threads regarding the inclusion of service/duration times at nodes, and I am left with a fundamental question. 

I know that the conventional method of including service times involves adding them to your transit callback so that the time to go between from_node -> to_node is given as: `transitTime[from_node][to_node] = serviceTime[from_node] + travelTime[from_node][to_node]`. However, I do not know if this accounts for the relative shift of time windows. 

A simple example that only involves Node A and Node B :

Let's say we have a time window on node B such that it must be completed within the range [10, 13]. Let's say it takes 10 time units to arrive from Node A to Node B, and Node B has a service time of 5 units. Logically, the latest that a vehicle can leave from Node A is at time 3, since then it will arrive at time 13. Then it arrives at time 13, waits for 5 units (service time), and then it is free to leave (here, the assumption is that time windows apply to arrival times). 

However, if the transit callback includes the service time, now it is impossible for the vehicle to have a feasible path from node A to node B, since the solver thinks it takes 15 units to reach from node A to node B. Therefore, it is impossible for the vehicle to reach Node B at any point between the arrival time window range of [10,13], even though the transit time is really only 10 units. 

Am I misunderstanding fundamentally how the time windows work? Or does this somehow already get taken care of? Issues [https://github.com/google/or-tools/issues/765](765) and [https://github.com/google/or-tools/issues/498](498) supposedly discuss service times, but they do not seem to address my confusion.

Thanks!

**Comment by Mizux** [comment]
Seems you are mixing ServiceTime(A) and ServiceTime(B) in you example. Since TW is a arrival time transit use serviceTime of the **from_node** not the **to_node**.

In you example `service_time(A)` is 0. So you have
`CumulVar(B) = CumulVar(A) + Service_time(A) + Travel_time(A, B)`
so if the TW at B is `[10, 13]` aka CumulVar(B) is in range [10, 13]. So if you leave A at 3 aka CumulVar(A) == 3, then you have it.

And yes, if `ServiceTime(A) > 3` there is no solution since `ServiceTime(A) + travelTime(A, B) > 13`...
